### PR TITLE
Handle upload error inside async uploader script

### PR DESCRIPTION
**Pull request by yossipapi**

### CODE REVIEW

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 1]
```diff
@@ -1,40 +1,49 @@
...
+//Read config file
 $configFilePath = $argv[1];
 $config = parse_ini_file($configFilePath);
 
-//Load configuration
-$files = glob($config['dirName'] . '/*.xml');
-if (!count($files))
+//Create REquired Directory For Script Execution
+if(!createDirIfNotExists($config['dirName'] . "/error/"))
 {
```
Comment (kalturaguy) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: Please add log here

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 2]
```diff
@@ -1,40 +1,49 @@
...
-logMsg(count ($files) . " files found.");
-
-$inProgress = array();
-if (file_exists('in_progress'))
+	
+if(!createDirIfNotExists($config['dirName'] . "/complete/"))
 {
-	$content = file_get_contents('in_progress');
-	$inProgress = explode('\n', $content);
+	die();
```
Comment (kalturaguy) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: Please add log here

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 3]
```diff
@@ -139,22 +154,126 @@ function getContentResource ($filepath, KalturaLiveStreamEntry $liveEntry, $work
...
 	
-	return null;
+	$e = new Exception("Error getting content resource");
+	logException($e);
+	
+}
+
+function handleException($xmlFilePath, $filePathOnDisc, $e, &$retryAttempts)
+{
+	$retryAttempts++;
```
Comment (kalturaguy) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: Please add log here

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 4]
```diff
@@ -139,22 +154,126 @@ function getContentResource ($filepath, KalturaLiveStreamEntry $liveEntry, $work
...
+    	$error = error_get_last();
+    	logMsg("Failed to create error dir with error " . implode(" __ ", $error));
+    }
+
+	return $res;
+}
+
+function moveFile($currFilePath, $newFilePath)
+{
+	rename($currFilePath, $newFilePath);
```
Comment (kalturaguy) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: Log...

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 5]
```diff
@@ -1,73 +1,81 @@
...
+			catch(Exception $e) {
+				$fileStatus = handleException($e, $f, strval($taskXml->filepath), strval($taskXml->entryId));
+			}
 		}
 		
 	}
-	$complete[] = $f;
-	file_put_contents('complete', implode('\n', $complete));
+	
+	if($filaeStatus == FILE_STATUS_DONE)
```
Comment (tan-tan-kanarek) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: $filaeStatus should be $fileStatus

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 6]
```diff
@@ -1,73 +1,81 @@
...
-	}
-	
-	$inProgress[] = $f;
-	file_put_contents('in_progress', implode('\n', $inProgress));
-	
-	$handled = false;
-	while (!$handled)
+{	
+	$fileStatus = FILE_STATUS_IN_PROGRESS;
+	while ($fileStatus == FILE_STATUS_IN_PROGRESS)
```
Comment (tan-tan-kanarek) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: You don't need the while anymore, the results of handleUploadXMLResource is always FILE_STATUS_DONE, which also means that you don't have to check the returned value, or even to return a value.

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 7]
```diff
@@ -126,35 +126,122 @@ function getContentResource ($filepath, KalturaLiveStreamEntry $liveEntry, $work
 		return $resource;
 	}
 	else {
-		try {
-			$client->startMultiRequest();
-			$client->uploadToken->add(new KalturaUploadToken());
+		$client->startMultiRequest();
+		$client->uploadToken->add(new KalturaUploadToken());
```
Comment (tan-tan-kanarek) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: Please add log before the API call.

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 8]
```diff
@@ -126,35 +126,122 @@ function getContentResource ($filepath, KalturaLiveStreamEntry $liveEntry, $work
...
-				if ($tokenResponse instanceof Exception) {
-			}
-				logMsg("Content resource creation error: " . $tokenResponse->getMessage() . "]");
-				return null;
+		$resource = new KalturaUploadedFileTokenResource();
+		$tokenResponse = $responses[1];
+		if ($tokenResponse instanceof KalturaUploadToken)
+			$resource->token = $tokenResponse->id;
+		else {
+			if ($tokenResponse instanceof Exception) {
```
Comment (tan-tan-kanarek) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: $tokenResponse will never be instance of Exception. Check if it's array that contains code key.

File: AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php [hunk 9]
```diff
@@ -1,73 +1,81 @@
...
 	{
 		$taskXml = new SimpleXMLElement(file_get_contents($f));
 		if ($taskXml->getName() == 'upload')
 		{
-			$handled = handleUploadXMLResource($taskXml, $client);	
+			try {
+				$fileStatus = handleUploadXMLResource($taskXml, $client);
+			}
+			catch(Exception $e) {
+				$fileStatus = handleException($e, $f, strval($taskXml->filepath), strval($taskXml->entryId));
```
Comment (tan-tan-kanarek) [inline AsyncMediaServerProcessClientApp/asyncMediaServerProcessScript.php]: Why won't you catch KalturaClientException in a different catch block and use a different method to handle the retry?

**Pull request closed without merging**